lib/ex_aws/auth/presigned_posts/policy.ex

```diff
@@ -0,0 +1,34 @@
+defmodule ExAws.Auth.PresignedPosts.Policy do
+  @moduledoc false
+  import ExAws.Auth.Utils
+
+  defstruct [expiration: nil, conditions: []]
+
+  def new do
+    %__MODULE__{}
+  end
+
+  def set_expiration(policy, datetime) do
+    %__MODULE__{policy | expiration: datetime}
+  end
+
+  def add_condition(policy = %__MODULE__{conditions: conditions}, condition) do
+    new_conditions =
+      conditions
+      |> Enum.reverse
+      |> List.insert_at(0, condition)
+      |> Enum.reverse
+    %__MODULE__{policy | conditions: new_conditions}
```

Comment:
Here's my problem with the approach here. This is an internal module; we have `@moduledoc false`. This is a relatively enormous amount of logic for what, as far as I could tell, could just be a list literal.

It is really very hard to follow what is going on. Every time I go to some function I have to jump over to 3 different modules just to figure out what's going on. It does not feel functional.

For example consider the simple line: `|> PresignedPost.set_bucket(bucket_name)`

To figure out what this does I have to read
```
def set_bucket(presigned_post, bucket_name) do
  put_policy_condition(presigned_post, %{"bucket" => bucket_name})
end
```
but this doesn't actually tell me anything so I go to `put_policy_condition`
```
def put_policy_condition(%{policy: policy} = presigned_post, condition) do
  policy = Policy.add_condition(policy, condition)
  %{presigned_post | policy: policy}
end
```
this still doesnt' tell me anything so I go to `Policy.add_condition`
```
def add_condition(policy = %__MODULE__{conditions: conditions}, condition) do
  new_conditions =
    conditions
    |> Enum.reverse
    |> List.insert_at(0, condition)
    |> Enum.reverse
  %__MODULE__{policy | conditions: new_conditions}
end
```
It's here I finally learn that we're just sticking the value at the end of a list somewhere.

As best I can tell this could all be replaced with just
```
policy = [
  ... stuff before
  %{"bucket" => bucket}
  ... stuff after
]
```

Your test cases actually illustrate this well:
```
test "adding multiple entries retains order" do
  policy =
    Policy.new
    |> Policy.add_condition(%{"acl" => "public-read"})
    |> Policy.add_condition(["starts_with", "$key", "something/"])
    |> Policy.add_condition(%{"Content-Type" => "image/jpeg"})

  assert policy.conditions == [
    %{"acl" => "public-read"},
    ["starts_with", "$key", "something/"],
    %{"Content-Type" => "image/jpeg"},
  ]
end
```
The actual `assert` stuff is shorter AND simpler than the API itself.